Multiple inputs and downloading of precomputed files per-input

jameshadfield opened the pull request:
WIP. PR up for discussion. Most probably doesn't work, but the DAGs are indicative of the direction. Only the initial download rules (sequences+metadata) and the `download_filtered` rule have been touched -- the others (`prefiltered`, `aligned` etc) are todos.

--- 

This PR begins to deprecate our former syntax for defining inputs:

```yaml
sequences: data/sequences.fasta
metadata: data/metadata.tsv
```
This will continue to work, but (a) does not allow multiple input origins and (b) files _must_ be present locally, i.e. cannot be downloaded. One such example of this is `snakemake --cores 4 --profile my_profiles/example -f auspice/ncov_north-america_usa_washington_king-county.json --dag | dot -Tpdf > dag_subsample_single.pdf`.

This PR introduces newer syntax, where inputs are defined per-origin, allow remote resources to be used, and allow downloading precomputed assets. Here are three examples for multiple inputs:

```yaml
### INPUT SET 1: Everything running locally, from local starting files
inputs:
  aus: 
    metadata: "data/example_metadata_aus.tsv"
    sequences: "data/example_sequences_aus.fasta"
  worldwide:
    metadata: "data/example_metadata_worldwide.tsv"
    sequences: "data/example_sequences_worldwide.fasta"

### INPUT SET 2: Gisaid inputs downloaded from S3. Everything running locally.
inputs:
  aus: #
    metadata: "data/example_metadata_aus.tsv"
    sequences: "data/example_sequences_aus.fasta"
  gisaid:
    metadata: "s3://nextstrain-ncov-private/metadata.tsv.gz"
    sequences: "s3://nextstrain-ncov-private/sequences.fasta.gz"

### INPUT SET 3: Gisaid pre-computed filtered alignment downloaded from S3
inputs:
  aus: # local files specified...
    metadata: "data/example_metadata_aus.tsv"
    sequences: "data/example_sequences_aus.fasta"
  gisaid:
    metadata: "s3://nextstrain-ncov-private/metadata.tsv.gz"
    filtered: "s3://nextstrain-ncov-private/filtered.fasta.xz"
```

These three input sets correspond to the following DAGs:

![image](https://user-images.githubusercontent.com/8350992/107469816-5695eb00-6bcf-11eb-9e13-c6a86ad27342.png)



### Nextstrain pipeline

The nextstrain config would now contain one of the following two starting points:

```yaml
# align GISAID data locally
inputs:
  gisaid:
    metadata: "s3://nextstrain-ncov-private/metadata.tsv.gz"
    sequences: "s3://nextstrain-ncov-private/sequences.fasta.gz"

# use precomputed GISAID filtered alignment
inputs:
  gisaid:
    metadata: "s3://nextstrain-ncov-private/metadata.tsv.gz"
    filtered: "s3://nextstrain-ncov-private/filtered.fasta.xz"
```

which correspond to the following DAGs:

![image](https://user-images.githubusercontent.com/8350992/107470034-a2e12b00-6bcf-11eb-8a87-c1eb0a87add9.png)

trvrb commented:
I really like this. Very small comment that I'd expect `filtered` to just be filtered and not yet aligned, just because we usually go `augur filter` -> `augur align`. I'd be less surprised if this was called `aligned` instead of `filtered`.

tsibley commented:
@jameshadfield Following up on the remote file support in Snakemake that I mentioned earlier today, see [this example](https://github.com/nextstrain/ncov/compare/remote-files) of how we could leverage that support make local paths and remote URLs interchangeable for the rules without needing to special-case via wildcards and input functions.

jameshadfield commented:
Quick update of the status of this PR: most use-cases should work, but some more testing is needed. 

Todo:
- [ ] Improve subsampling declarations [see John's review on 549](https://github.com/nextstrain/ncov/pull/549#pullrequestreview-588882659)
- [ ] Make sure the functions in `export_for_nextstrain.smk` work (I suspect not)
- [ ] Update documentation
- [ ] Improve the distinction between `origin`s with and without a leading `_`, [see John's review on 549](https://github.com/nextstrain/ncov/pull/549#pullrequestreview-588882659)
- [ ] Some more testing
- [ ] Try using Snakemake's built in downloading, but most probably that will be left for a subsequent PR

jameshadfield commented:
This PR is now ready for re-review and merging 😄 The tutorial has been updated to explain the various functionality introduced here:
* Multiple (starting) inputs
* Declarative syntax if the input files reside on S3
* Ability to declare intermediate starting files which reside on S3, e.g. starting from a filtered alignment (see below) 

### Testing
I've tested this via the following situations
* `example tutorial`
* `example_multiple_inputs` tutorial (added here)
* core nextstrain pipeline (starting from unaligned sequences). Job ID `666f9e83-333c-43b1-aeba-ba0c3e529217` if you want to check this out!
* Uploading a filtered alignment to our private bucket via `rule upload` (see below)
* core nextstrain pipeline starting from that filtered alignment on s3. Job ID `637981dd-8b6f-4377-bcd6-ed1e2b5b279b` if you want to check this out!


### Changes to nextstrain core pipeline

cc @MoiraZuber @eharkins 

The starting files are now updated to use the `config["inputs"]` syntax, which makes explicit that these files come from GISAID. This is reflected in some of the produced filenames - e.g. `results/aligned.fasta` is now `results/aligned_gisaid.fasta`. See comments added in `nextstrain/builds.yaml`, including instructions on how to start the pipeline from a preprocessed filtered alignment.

### Changes to subsampling declarations

cc @huddlej 

I've made minor changes to our rules to make `group_by` optional (0a615aa) which cleans things up a bit. I didn't pursue adding a more expressive syntax for specifying "take all seqs from this input", for instance:
```yaml
# Current syntax
    allFromAus:
      exclude: "--exclude-where 'aus!=yes'" # subset to sequences from input `aus`
# Better syntax
    allFromAus:
      use_all_from_input: aus
```

But this should be relatively simple in the future (e.g. have `_get_subsampling_settings` check that `exclude` and `use_all_from_input` aren't both defined, and then have `_get_specific_subsampling_setting` translate `use_all_from_input: aus` to `--exclude-where 'aus!=yes'`). Thoughts?

### Uploading of preprocessed files

cc @rneher @emmahodcroft 

The changes mean that preprocessed files now include the origin in the filename. I've updated the `rule upload` accordingly and there should be no changes to how it's run. So far I've only uploaded a `filtered_gisaid.fasta.xz` alignment to the bucket. The rule uses new config params `S3_DST_BUCKET`, `S3_DST_COMPRESSION` and `S3_DST_ORIGINS`. Happy to take an alternate approach here if there's a better one!

### Starting a build from preprocessed files

cc @trvrb

See the comments at the top of `nextstrain_profiles/nextstrain/builds.yaml` for examples of how to start the pipeline from a variety of preprocessed starting points. I've written a brief "how to" in the tutorial, but stressed that the names we use may change shortly.


### Changes to `origin` and variables within the snakemake file

cc @huddlej 

I've added a helper function `_trim_origin` insted of lots of `wildcards.origin[1:]` calls, and renamed variables where both the wildcards origin and the trimmed origin are used simultaneously, which I help makes things easier to read.

jameshadfield commented:
TravisCI is currently failing at steps which download from S3, e.g.

```sh
# Error in rule download_metadata:
aws s3 cp s3://nextstrain-ncov-private/metadata.tsv.gz - | gunzip -cq >data/downloaded_gisaid.tsv
```

However these rules work fine for me locally. @tsibley is this a credential issue which I'm missing by chance?

tsibley commented:
@jameshadfield Ah, I didn't configure the Travis CI jobs for this repo and don't know what (if any) AWS creds are provided to it. I suspect that it was not setup with using AWS in mind.

rneher commented:
I am testing this now. We could also remove all these `prefilter` and `refilter` steps. I think we only needed them when we were afraid messed up sequences would screw up the alignment. but with everything being reference aligned, that shouldn't make a difference.

huddlej commented on scripts/combine_metadata.py:
```diff
@@ -0,0 +1,80 @@
+
+if __name__ == '__main__':
+    args = parse_args()
+    assert(len(args.metadata)==len(args.origins))
+    assert(len(args.origins)>1)
```
Could we convert these assertions to user-friendly error messages printed to stderr when either test fails?

jameshadfield replied:
Yup! a16d89f

huddlej commented on scripts/combine_metadata.py:
```diff
@@ -0,0 +1,80 @@
+    # SUMMARISE INPUT METADATA
+    print(f"Parsed {len(metadata)} metadata TSVs")
+    for m in metadata:
+        print(f"\t{m['origin']} ({m['fname']}): {len(m['data'].keys())} strains x {len(m['columns'])} columns")
```
I like that I got this extra feedback on the command line.

huddlej reviewed (commented):
This looks really good, @jameshadfield. Although this runs for me locally and the tutorial is really clear, I'm less confident about reviewing the AWS/S3 changes (I'm the one who setup Travis CI without any awareness of AWS credentials, so clearly not an S3 expert... ;) ) or saying that this could be merged without disrupting current builds.

I do have a thought about how to implement the `origin` wildcard in a way that doesn't require any special parsing, but I haven't had a chance to test it yet. I don't think that design decision should hold up this PR though.

huddlej commented on nextstrain_profiles/nextstrain/builds.yaml:
```diff
@@ -2,10 +2,36 @@
+# E.g. if you uncomment `filtered` then the pipeline
+# will start by downloading that file and proceeding straight to
+# subsampling
+inputs:
+  gisaid:
+    metadata: "s3://nextstrain-ncov-private/metadata.tsv.gz"
+    sequences: "s3://nextstrain-ncov-private/sequences.fasta.gz"
```
@jameshadfield @tsibley This change is what's causing the CI tests to fail. The CI test was setup to run a Nextstrain global build on the "example data" from the repo without downloading any external data. It does this by overriding the `sequences` and `metadata` values of the config on the command line:

```
nextstrain build . all_regions -j 2 --profile nextstrain_profiles/nextstrain --config sequences=data/example_sequences.fasta metadata=data/example_metadata.tsv active_builds=global
```

But now that there are nested inputs here with different names, we'll need to use a different approach to testing. One fix would be to allow overriding `inputs` to an empty string from the command line like so:

```
nextstrain build . all_regions -j 2 --profile nextstrain_profiles/nextstrain --config sequences=data/example_sequences.fasta metadata=data/example_metadata.tsv active_builds=global inputs=
```

To do this, we would need to change a couple of Snakemake lines from:

```
if "inputs" not in config:
```

To:

```
if not config.get("inputs"):
```

These minimal changes make the CI command work for me again locally. The question remains whether we want to continue supporting the `sequences` and `metadata` keys in the future even for this minimal test case.

huddlej replied:
Here is the patch for all the changes I needed to make:

```diff
diff --git a/workflow/snakemake_rules/common.smk b/workflow/snakemake_rules/common.smk
index 4618e04..fe52456 100644
--- a/workflow/snakemake_rules/common.smk
+++ b/workflow/snakemake_rules/common.smk
@@ -49,7 +49,7 @@ def _get_path_for_input(stage, origin_wildcard):
     if not origin_wildcard:
         # No origin wildcards => deprecated single inputs (e.g. `config["sequences"]`) which cannot
         # be downloaded from remote resources
-        if "inputs" in config:
+        if config.get("inputs"):
             raise Exception("ERROR: empty origin wildcard but config defines 'inputs`")
         path_or_url = config[stage] if stage in ["metadata", "sequences"] else ""
         remote = False
@@ -97,14 +97,14 @@ def _get_unified_metadata(wildcards):
     then that file is returned. Else "results/combined_metadata.tsv" is returned
     which will run the `combine_input_metadata` rule to make it.
     """
-    if "inputs" not in config:
+    if not config.get("inputs"):
         return config["metadata"]
     if len(list(config["inputs"].keys()))==1:
         return _get_path_for_input("metadata", "_"+list(config["inputs"].keys())[0])
     return "results/combined_metadata.tsv"

 def _get_unified_alignment(wildcards):
-    if "inputs" not in config:
+    if not config.get("inputs"):
         return "results/filtered.fasta"
     if len(list(config["inputs"].keys()))==1:
         return _get_path_for_input("filtered", "_"+list(config["inputs"].keys())[0])
```

jameshadfield replied:
Resolved by 73c3455 - thanks John!

jameshadfield commented on docs/multiple_inputs.md:
```diff
@@ -0,0 +1,251 @@
+In the case of **conflicts**, the order of the entries in the YAML matters, with the last value being used.
+
+Finally, we use one-hot encoding to express the origin of each row of metadata.
+This means that **extra columns** will be added for each input (e.g. `aus` and `worldwide`), with values of `"yes"` or `""`, representing which samples are contained in each set of sequences.
```
Ha! [yes, yes , yes](https://www.youtube.com/watch?v=0bYY8m1Lb2I) is rather confusing!

jameshadfield replied:
Shifted to explicit "yes"/"no" values in a16d89f (the display problems are part of auspice).

rneher commented:
It would be really helpful if we could get this merged soon as this is touching so many parts of the code and any other developments will cause conflicts...

the PR works well for me on sciCore.

jameshadfield commented on docs/multiple_inputs.md:
```diff
@@ -0,0 +1,251 @@
+
+The resulting JSON can be dropped onto [auspice.us](https://auspice.us) for visualization.
+
+> P.S. If you want to see the DAG to help understand the pipeline steps, you can run
```
Resolved by decbfd7

jameshadfield commented on docs/multiple_inputs.md:
```diff
@@ -0,0 +1,251 @@
+
+> You may use `.xz` or `.gz` compression - we automatically infer this from the filename suffix.
+
+### Can I start from intermediate files stored remotely?
```
Good call. Resolved by decbfd7.

jameshadfield commented on scripts/combine_metadata.py:
```diff
@@ -0,0 +1,80 @@
+        description="Custom script to combine metadata files",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+    parser.add_argument('--metadata', required=True, nargs='+', metavar="TSV", help="Metadata files")
```
Good call! Updated the help message in a16d89f, and now also explain the one-hot encoding we use.

jameshadfield commented on Snakefile:
```diff
@@ -1,6 +1,6 @@
 import copy
 from datetime import date
-import os
+import os, sys
```
No problem -- done in 99bd0f5

jameshadfield commented on docs/multiple_inputs.md:
```diff
@@ -0,0 +1,251 @@
+    sequences: "data/example_sequences_aus.fasta"
+  worldwide:
+    metadata: "data/example_metadata_worldwide.tsv"
+    sequences: "data/example_sequences_worldwide.fasta"
```
Shifted to this approach in 99bd0f5, and updated the schema to enforce the "name" and "metadata" properties. (As an aside, we could be using the schema a lot more 😉). Internally, we convert this to an `OrderedDict` straightaway so that we can still access inputs by "name" keyword throughout the rules.

huddlej replied:
Awesome! Thanks, @jameshadfield!

jameshadfield commented:
Tests & example builds pass, and given the above comments I think I'm going to merge (along with #554), and then rebase #555 onto the (new) master branch & test locally.

tsibley commented:
Hmm, I didn't notice before, but these changes:

> This is reflected in some of the produced filenames - e.g. `results/aligned.fasta` is now `results/aligned_gisaid.fasta`.

> The changes mean that preprocessed files now include the origin in the filename. I've updated the rule upload accordingly and there should be no changes to how it's run. So far I've only uploaded a `filtered_gisaid.fasta.xz` alignment to the bucket.

have implications for S3 access policies on the private bucket. I've updated policies (`NextstrainNcovPrivateUploader` and `NextstrainNcovTrialRunner`) to broaden access for our automation instead of continuing to use more fine-grained access control (which anticipated more segregated access to the bucket than ever came to pass and was more manageable with our early-pandemic automation).
